### ISSUE TITLE
SwingObservable subscribe/unsubscribe: Who has to ensure it's done on the EDT?

### ISSUE THREAD

**Issue by samuelgruetter** [ReactiveX/RxSwing#3]
When updating my [Swing RxScala examples](https://github.com/samuelgruetter/rx-playground) to RxScala 0.20.4, I suddenly got many `java.lang.IllegalStateException: Need to run in the event dispatch thread`. I found out that the changes which broke my apps came from https://github.com/ReactiveX/RxJava/pull/883, and I also read the discussion in https://github.com/ReactiveX/RxJava/pull/869. There, I learnt that not only modifications of the UI, but also modifications of the listeners (addXxxListener/removeXxxListener) have to run on the event dispatch thread. I had only respected the first, but not the second of these constraints. Knowing this, it was easy to fix my apps. Problem solved, end of the story ;-)

But when I looked at the new SwingObservable, I found an asymmetry: When I subscribe to a Swing Observable, I have to ensure myself that I'm on the event dispatch thread, otherwise an IllegalStateException is thrown (see [this line](https://github.com/ReactiveX/RxSwing/blob/0.x/src/main/java/rx/swing/sources/KeyEventSource.java#L43)).
However, when I unsubscribe, I can do so from any thread, and the Swing Observable ensures that the unsubscription is run on the event dispatch thread (see [this line](https://github.com/ReactiveX/RxSwing/blob/0.x/src/main/java/rx/swing/sources/KeyEventSource.java#L62)).

This seems to be an inconsistent way of distributing the responsabilities between the library and the library users. @zsxwing do you agree? Or is there a good reason for this asymmetry that I'm missing?

**Comment by zsxwing** [comment]
Alternatively, we can call `subscribeOn(SwingScheduler. getInstance())` and remove the assert. I don't remember why I didn't choose `subscribeOn`. What do you think?

**Comment by samuelgruetter** [comment]
I think we can summarize the problem as follows: We have to ensure that subscribing and unsubscribing happens on the EDT, and both for subscribing und unsubscribing, there are the same four ways of achieving this:
1.    Require in the documentation that the library _user_ has to ensure that we're on the EDT, but never check it. That's what the Swing library does.
2.    Require in the documentation that the library _user_ has to ensure that we're on the EDT, check it using `SwingUtilities.isEventDispatchThread()`, and throw an exception if it returns `false`. That's what currently happens when you subscribe to a Swing Observable, and it hinted me at an issue in my apps: I learnt that not only GUI modifications, but also the GUI construction has to be done on the EDT (see also [this stackoverflow question](http://stackoverflow.com/questions/491323/is-it-safe-to-construct-swing-awt-widgets-not-on-the-event-dispatch-thread)).
3.    The library takes care of ensuring that we're on the EDT, by using `subscribeOn(SwingScheduler.getInstance())` and `unsubscribeOn(SwingScheduler.getInstance())`. That's simple for the library implementors as well as for the library users.
4.    The library takes care of ensuring that we're on the EDT, by using something like `SwingSubscriptions.unsubscribeInEventDispatchThread`, which only switches thread if `SwingUtilities.isEventDispatchThread()` returns `false`. This might (or might not, I don't know) be slightly more efficient than 3).

As you can see, each of these four options has an advantage, so I'm not (yet) sure which one to pick.

/cc @headinthebox @jmhofer

**Comment by zsxwing** [comment]
After carefully consideration, I vote 2 because it can teach the library user don't touch the Swing component out of the EDT if they don't know the Swing rules. For `unsubscribeOn`, `SwingSubscriptions.unsubscribeInEventDispatchThread` and `unsubscribeOn(SwingScheduler.getInstance())` are good to me.

**Comment by headinthebox** [comment]
This is why `subscribeOn` was designed (and only for that purpose) that _both_ subscription and unsubscription happen on the specified thread (for a while the implementation of `subscribeOn` did not unsubscribe on the correct thread).

So the preferred solution would be the first half of 4.

**Comment by epb-644** [comment]
I know it's been awhile, but does anyone still feel strongly about this issue? 

My personal feeling is that our current behavior is fine. I think Swing users will be aware enough of threading to call `subscribeOn` when necessary. And the `assertEventDispatchThread()` is a nice hand-slap if they aren't.

Thoughts? Is it still worth pursuing any changes?

**Comment by mikebaum** [comment]
Just my two cents. I would tend to agree with @headinthebox. Aside from what has already been mentioned, I think it's safe to say that anyone who uses this library is probably already aware of the need to be on the EDT. Thus, having this detail taken care of by the library would be appreciated.

I would be okay with Option 3 above. However If we go with option 4, I guess we could factor out the content of the method: `unsubscribeInEventDispatchThread()` to be re-used for both the `subscribe` and `unsubscribe` actions. Perhaps the following method would make sense:

```
public static void invokeNowOrLaterOnEDT( Action0 action ) {
    if (SwingUtilities.isEventDispatchThread()) {
        action.call();
    } else {
        final Worker inner = SwingScheduler.getInstance().createWorker();
        inner.schedule(new Action0() {
            @Override
            public void call() {
                action.call();
                inner.unsubscribe();
            }
        });
    }
}
```

**Comment by samuelgruetter** [comment]
That makes sense. But there's one thing I don't quite get: Why is the `else` branch so complicated? Couldn't we just use `SwingUtilities.invokeLater(Runnable)`?

**Comment by mikebaum** [comment]
@samuelgruetter, I was going to suggest that, but I wasn't sure if there was some kind of advantage to using the SwingScheduler. I suppose in this case we would not ever need to use the advanced features of a `Scheduler` so `SwingUtilities.invokeLater(Runnable)` would be fine. So the updated method would be:

```
public static void invokeNowOrLaterOnEDT(final Action0 action) {
    if (SwingUtilities.isEventDispatchThread()) {
        action.call();
    } else {
        SwingUtilities.invokeLater(new Runnable() {
            @Override
            public void run() {
                action.call();
        } } );
    }
}
```

Of course changing the method signature to take a `Runnable` rather than an `Action0` would further simplify the method. With this change, we'd need to update all the EventSource objects appropriately. I could prepare a patch for this, if you like?

**Comment by samuelgruetter** [comment]
Don't we have to update all the usages in the event sources anyways? Because `invokeNowOrLaterOnEDT` returns void, whereas `unsubscribeInEventDispatchThread` returned `Subscription`.

now:

``` java
subscriber.add(SwingSubscriptions.unsubscribeInEventDispatchThread(new Action0() {
   @Override
   public void call() {
      component.removeComponentListener(listener);
   }
}));
```

new:

``` java
subscriber.add(new Action0() {
   @Override
   public void call() {
      invokeNowOrLaterOnEDT(...)
   }
});
```

With this in mind, we could indeed give a `Runnable` to `invokeNowOrLaterOnEDT`.

But I'd first like to hear @zsxwing 's opinion, who wrote the original `SwingSubscriptions.unsubscribeInEventDispatchThread` (commit https://github.com/ReactiveX/RxSwing/commit/c8867096f285a4ad9fa25b620a5f493fb2d8a85f).

**Comment by zsxwing** [comment]
`invokeNowOrLaterOnEDT(Runnable)` looks good. However, because `Subscriber.add(Action0)` does not exist in RxJava. The new code should be:

``` java
subscriber.add(Subscriptions.create(new Action0() {
   @Override
   public void call() {
      invokeNowOrLaterOnEDT(...)
   }
}));
```

**Comment by mikebaum** [comment]
Okay, super. So I'll go ahead and prepare a patch.

I'm temped to make the method `invokeNowOrLaterOnEDT` package protected. It's a nice helper for setting up EventSources but it's use should probably be discouraged in favor of the `SwingScheduler` elsewhere. The risk of using this method more generally is that it could cause events to be processed out of order. Anyway, maybe that's too much paranoia.

Let me know if you have a preference one way or the other.

**Comment by headinthebox** [comment]
I think this goes totally against the spirit of Rx where all issues wrt to threading are handled _explicitly_ via schedulers and `subscribeOn` to make sure the subscription and unsubscription happens on the right thread, and `observeOn` that ensures the observations happen on the right thread.

Historical note, the same UI thread issues happen in Windows, where Rx was born. We tried to make threading issues implicit which led to subtle bugs that were impossible to track, but in the we came up with the idea of schedulers and life was good.

By not using schedulers you close the door to use the test scheduler for example.

Anyway, that's just my 2 cents.

**Comment by mikebaum** [comment]
@headinthebox, you mentioned earlier that providing a `Scheduler` via `subscribeOn` would be sufficient for both the subscribe and unsubscribe actions. Is this really the case? I see that there is also an `unsubscribeOn` method. 

Anyway I would be okay with using `subscribeOn` and providing the SwingScheduler as the default `Scheduler`. I believe this is a good compromise. In this way callers could still change the `Scheduler` as they wish.

**Comment by headinthebox** [comment]
The only goal of `subscribeOn` is to make sure the subscription and unsubscription happen on the specified scheduler. Any implementation that does not guarantee that has a bug.

**Comment by zsxwing** [comment]
> The only goal of subscribeOn is to make sure the subscription and unsubscription happen on the specified scheduler. Any implementation that does not guarantee that has a bug.

`unsubscribeOn` can not guarantee that in RxJava. Here is an example:

``` Java
        Observable.create(new OnSubscribe<Integer>() {

            @Override
            public void call(final Subscriber<? super Integer> t1) {
                final Subscription s = Subscriptions.create(new Action0() {

                    @Override
                    public void call() {
                        System.out.println(Thread.currentThread().getName());
                    }
                });
                t1.add(s);
                t1.onNext(1);
                t1.onCompleted();
            }
        }).take(1).observeOn(Schedulers.io()).unsubscribeOn(Schedulers.newThread()).subscribe();
```

The above codes will output `main` rather than `RxNewThreadScheduler-1`.

The problem here is that `take(1)` does not know `unsubscribeOn` in the downstream and simply call `unsubscribe` to trigger the upstream `Subscription`.

**Comment by headinthebox** [comment]
The general pattern is `xs.subscribeOn(...).ys.observeOn(...).subscribe(...)`, i.e. one as early as possible, the other as late as possible.

**Comment by mikebaum** [comment]
Confirmed... running the following code yields the desired results:

```
        Observable.create(new OnSubscribe<Integer>() {

            @Override
            public void call(final Subscriber<? super Integer> t1) {
                System.out.println("Subscribing on: " + Thread.currentThread().getName());
                final Subscription s = Subscriptions.create(new Action0() {

                    @Override
                    public void call() {
                        System.out.println("Unsubscribing on: " + Thread.currentThread().getName());
                    }
                });
                t1.add(s);
                t1.onNext(1);
                t1.onCompleted();
            }
        }).subscribeOn(SwingScheduler.getInstance()).take(1).observeOn(Schedulers.io()).toBlocking().first();
```

output:

```
Subscribing on: AWT-EventQueue-0
Unsubscribing on: AWT-EventQueue-0
```

With that in mind why don't we just update all the EventSources to tack on a call to `subscribeOn(SwingScheduler.getInstance())` at the end. E.g `ItemEventSource.fromItemEventsOf()` would become:

```
public static Observable<ItemEvent> fromItemEventsOf(final ItemSelectable itemSelectable) {
        return Observable.create(new OnSubscribe<ItemEvent>() {
            @Override
            public void call(final Subscriber<? super ItemEvent> subscriber) {
                final ItemListener listener = new ItemListener() {
                    @Override
                    public void itemStateChanged(ItemEvent event) {
                        subscriber.onNext(event);
                    }
                };
                itemSelectable.addItemListener(listener);
                subscriber.add(Subscriptions.create( new Action0() {
                    @Override
                    public void call() {
                        itemSelectable.removeItemListener(listener);
                    }
                }));
            }
        }).subscribeOn( SwingScheduler.getInstance() );
    }
```

**Comment by headinthebox** [comment]
That's how things were meant to be ;-)

**Comment by epb-644** [comment]
So just to be clear: Do we have a consensus that every EventSource should itself call `subscribeOn( SwingScheduler.getInstance())`? 

I think we should move forward, unless there are strong disagreements.

Personally, I would still just throw exceptions for off-EDT subscriptions & unsubscriptions (option 2 from @samuelgruetter). But I'm biased -- I've dealt with too much code where the devs never bothered to understand Swing's threading policy, and I always wished their code would blow up to teach them a lesson. :)

That said, I think that implicitly subscribing & unsubscribing on EDT is an improvement. I'm on board.

Anybody totally against implicit `subscribeOn( SwingScheduler.getInstance())` calls?

**Comment by mikebaum** [comment]
My position hasn't changed, I'm for the addition of the implicit `subscribeOn( SwingScheduler.getInstance())`. If everyone agrees, who will make the change? I have offered, but if someone else feels like doing it, by all means.

**Comment by zsxwing** [comment]
I'm not against the `subscribeOn` idea. Don't forget to add `unsubscribeOn(SwingScheduler.getInstance())`. The `unsubscribeOn` issue won't happen in these APIs. My previous comment about `unsubscribeOn` is just to warn the potential issue, such as `take` is put before `unsubscribeOn`.

**Comment by mikebaum** [comment]
Okay no problem, even if the call to `unsubscribeOn` is technically not needed I do see the merit in being explicit.

**Comment by zsxwing** [comment]
> Okay no problem, even if the call to unsubscribeOn is technically not needed I do see the merit in being explicit.

`unsubscribeOn` is necessary because `subscribeOn` won't schedule the unsubscribe action. See https://github.com/ReactiveX/RxJava/pull/890

**Comment by epb-644** [comment]
@mikebaum Yeah if you want to do the pull request, that'd be great. (Btw, I promise I haven't forgotten about your other pull-request.)

**Comment by mikebaum** [comment]
@eddieburns55 No worries about the pull request. I know we'll all busy with day jobs. I'll take care of creating another pull request for the changes discussed here.

FYI, I'm somewhat eager to help complete the library as I'm pretty sure we're going to use it at my work. We develop a large Diagnostic Image Viewer which is nearly 100% Swing based.

**Comment by epb-644** [comment]
@mikebaum Thanks, and that's great to hear that you might use it at your work. Hopefully we can get all the features you need. We already use quite a bit at my job -- especially for mouse & keyboard events. I'll be interested to see what use-cases you find.

**Comment by epb-644** [comment]
Closed by pr #18
